api/src/main/java/org/apache/iceberg/DataFile.java

```diff
@@ -152,4 +153,16 @@ static StructType getType(StructType partitionType) {
    * are determined by these offsets. The returned list must be sorted in ascending order.
    */
   List<Long> splitOffsets();
+
+  /**
+   * @return sequential number of this data file if exists, null otherwise.
+   */
+  Long seqNum();
+
+  /**
+   * Assign a sequential number to data file.
+   *
+   * @param seqNum the sequential number to assign
+   */
+  void setSeqNum(long seqNum);
```

Comment:
This interface should not include a setter method. `DataFile` is effectively immutable.